RFC-0003-arrow-flight-connector.md

```diff
@@ -0,0 +1,122 @@
+
+1. Create a Flight Client using the host, port, ssl information of the Flight service. 
+    1. Maintain a connection pool of Flight clients.
+    2. Reuse a Flight client from the pool that is already connected to the same location as the new connection.
+    3. Evict connections from the pool using the Least Frequently Used strategy.
+2. Use the Flight Client to connect to the Flight service. 
+3. Create a FlightDescriptor using a command
+    1. The connector template will define an abstract class that returns a byte array.
+    1. The concrete implementation of this connector template will define a class that returns a byte array for the `FlightDescriptor.command` method. 
+    2. The concrete implementation will specify the necessary details for the database connection in the `command` byte array.
+4. Use the FlightDescriptor to obtain a FlightInfo object.
+    1. The FlightInfo object will be got from FlightClient by authenticating to the FlightClient using a bearer token.
+    2. The concrete implementation of the connector is responsible for generating or providing the bearer token.
+5. Use the FlightInfo object to obtain the list of endpoints we will be retrieving data from. 
+6. Use each endpoint to obtain a Ticket. 
+7. Use the Ticket to obtain a Stream. 
+8. Use the Stream to read data. 
+    1. Get VectorSchemaRoot from Flight stream
+    2. For each column in the result, build a Block from the FieldVector
+    3. Return a page with all the constructed Blocks.
+    4. Return new pages until all the Flight streams are read.
+
+#### Query execution using flight
+
+![Query execution using Flight](arrow-flight-connector/Query-execution-using-flight.png)
+
+### Low level design
+
+#### Connector splits in Presto for FlightInfo tickets
+
+Flight server can serve data using multiple tickets based on the connector type and query. Data can be fetched in parallel using the multiple flight tickets. 
+
+Flight descriptor will be created from a command of byte array. The concrete implementation is responsible for providing the byte array that differs from server to server. Flight Info will be got from the Flight Descriptor. Presto can create connector splits for each ticket from Flight Info so that data can be fetched in parallel by different worker nodes.
+
+![Connector splits in Presto](arrow-flight-connector/Connector-splits-in-Presto.png)
+
+#### Class diagram
+
+![Class diagram](arrow-flight-connector/Class-diagram.png)
+
+
+## Prestissimo Implementation
+A single Prestissimo connector will handle all concrete instances of the Java Arrow Flight connector. This is possible because the prestissimo connector does not have to deal with metadata fetch and discovery. Each split wraps a ticket and a list of locations, which is all[^1] that is required to fetch data from the Flight Service in an implementation-agnostic way.
```

Comment:
Its important to consider how the ticket and locations list will be serialized from Presto Java to Prestissimo. https://github.com/prestodb/presto/blob/master/presto-native-execution/presto_cpp/presto_protocol/presto_protocol.yml shows the other mappings. Please give some more info on how this will be generated.

Reply:
@ashkrisk Can you please check the comments on Prestissimo implementation

Reply:
The Java implementation looks something like this:
```java
class ArrowSplit {
  byte[] ticket;
  List<String> locations;
}
```
Jackson can serialize `byte[]` and `List<String>`. `byte[]` in particular will be serialized as a base64-encoded string.

The corresponding C++ implementation is quite similar:
```cpp
class FlightSplit {
  std::string ticket;
  std::vector<std::string> locations;
}
```
The only extra detail to take care of is that although the `base64` encoding/decoding is implicit in Java, it has to be handled explicitly in C++.

Reply:
Added a section with details https://github.com/prestodb/rfcs/blob/2d0740ac0b610057c95026879bd343d5ddc3cedc/RFC-0003-arrow-flight-connector.md?plain=1#L88